android/WALT/app/src/main/java/org/chromium/latency/walt/MidiFragment.java

```diff
@@ -0,0 +1,285 @@
+        MidiDeviceInfo[] infos = m.getDevices();
+        for(MidiDeviceInfo info : infos) {
+            String name = info.getProperties().getString(MidiDeviceInfo.PROPERTY_NAME);
+            logger.log("Found MIDI device named " + name);
+            if(TEENSY_MIDI_NAME.equals(name)) {
+                logger.log("^^^ using this device ^^^");
+                m.openDevice(info, new MidiManager.OnDeviceOpenedListener() {
+                            @Override
+                            public void onDeviceOpened(MidiDevice device) {
+                                if (device == null) {
+                                    logger.log("Error, unable to open MIDI device");
+                                } else {
+                                    logger.log("Opened MIDI device successfully!");
+                                    mMidiDevice = device;
+                                }
+                            }
+                        }, null);
+                break;
+            }
+        }
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        mTextView = (TextView) activity.findViewById(R.id.txt_box_midi);
+
+        // Register this fragment class as the listener for some button clicks
+        activity.findViewById(R.id.button_start_midi_in).setOnClickListener(this);
+        activity.findViewById(R.id.button_start_midi_out).setOnClickListener(this);
+
+        // mLogTextView.setMovementMethod(new ScrollingMovementMethod());
+        mTextView.setText(activity.logger.getLogText());
+        activity.logger.broadcastManager.registerReceiver(mLogReceiver,
+                new IntentFilter(activity.logger.LOG_INTENT));
+
+    }
+
+    @Override
+    public void onPause() {
+        logger.broadcastManager.unregisterReceiver(mLogReceiver);
+        super.onPause();
+    }
+
+    @Override
+    public void onClick(View v) {
+        if(mMidiDevice == null) {
+            logger.log("Wait until the MIDI devices connects");
+        } else {
+            switch (v.getId()) {
+                case R.id.button_start_midi_in:
+                    setupMidiIn();
+                    activity.handler.postDelayed(requestNoteRunnable, noteDelay);
+                    break;
+                case R.id.button_start_midi_out:
+                    setupMidiOut();
+                    scheduleNote();
+                    activity.handler.postDelayed(cancelMidiOutRunnable, noteDelay + timeout);
+                    break;
+                default:
+                    break;
+            }
+        }
+    }
+
+    private BroadcastReceiver mLogReceiver = new BroadcastReceiver() {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            String msg = intent.getStringExtra("message");
+            mTextView.append(msg + "\n");
+        }
+    };
+
+    private void setupMidiOut() {
+        mInputPort = mMidiDevice.openInputPort(0);
```

Comment:
We need some comment somewhere about the "in" vs "out" confusion. Why *MidiOut functions use mInputPort and vice versa.